src/Codec/CBOR/Cuddle/CBOR/Validator.hs

```diff
@@ -0,0 +1,980 @@
+    Postlude PTAny -> pure ()
+    -- a = nil
+    Postlude PTNil -> pure ()
+    _ -> throwError (ValLeaf rule UnapplicableRule)
+  pure (ValLeaf rule ())
+
+--------------------------------------------------------------------------------
+-- Bytes
+
+-- | Validating a byte sequence
+validateBytes ::
+  (MonadError (ValidationTree CDDLFail) m, MonadReader CDDL m) =>
+  BS.ByteString -> Rule -> m (ValidationTree ())
+validateBytes bs rule = do
+  getRule rule >>= \case
+    -- a = any
+    Postlude PTAny -> pure ()
+    -- a = bytes
+    Postlude PTBytes -> pure ()
+    -- a = h'123456'
+    Literal (Value (VBytes bs') _) -> check (ValLeaf rule) $ bs == bs'
+    -- a = foo .and bar
+    Control And tgt ctrl ->
+      validateBytes bs tgt >> void (withError (fmap (const $ ControlFail ctrl)) $ validateBytes bs ctrl)
+    -- a = foo .within bar
+    Control Within tgt ctrl ->
+      validateBytes bs tgt >> void (withError (fmap (const $ ControlFail ctrl)) $ validateBytes bs ctrl)
+    -- a = foo .ctrl bar
+    Control op tgt ctrl ->
+      validateBytes bs tgt
+        >> ifM (controlBytes bs op ctrl) (pure ()) (throwError (ValLeaf rule (ControlFail ctrl)))
+    -- a = foo / bar
+    Choice opts -> void $ validateChoice (validateBytes bs) rule opts
+    _ -> throwError (ValLeaf rule UnapplicableRule)
+  pure (ValLeaf rule ())
+
+-- | Controls for byte strings
+controlBytes :: forall m. MonadReader CDDL m => BS.ByteString -> CtlOp -> Rule -> m Bool
+controlBytes bs Size ctrl =
+  getRule ctrl >>= \case
+    Literal (Value (VUInt (fromIntegral -> sz)) _) -> pure $ BS.length bs == sz
+    Range low high bound ->
+      let i = BS.length bs
+       in ((,) <$> getRule low <*> getRule high) <&> \case
+            (Literal (Value (VUInt (fromIntegral -> n)) _), Literal (Value (VUInt (fromIntegral -> m)) _)) -> n <= i && range bound i m
+            (Literal (Value (VNInt (fromIntegral -> n)) _), Literal (Value (VUInt (fromIntegral -> m)) _)) -> -n <= i && range bound i m
+            (Literal (Value (VNInt (fromIntegral -> n)) _), Literal (Value (VNInt (fromIntegral -> m)) _)) -> -n <= i && range bound i (-m)
+controlBytes bs Bits ctrl = do
+  indices <-
+    getRule ctrl >>= \case
+      Literal (Value (VUInt i') _) -> pure [i']
+      Choice nodes -> getIndicesOfChoice nodes
+      Range ff tt incl -> getIndicesOfRange ff tt incl
+      Enum g -> getIndicesOfEnum g
+  bitsControlCheck (map fromIntegral indices)
+  where
+    bitsControlCheck :: [Int] -> m Bool
+    bitsControlCheck allowedBits =
+      let allowedSet = IS.fromList allowedBits
+          totalBits = BS.length bs * 8
+          isAllowedBit n =
+            let byteIndex = n `shiftR` 3
+                bitIndex = n .&. 7
+             in case BS.indexMaybe bs byteIndex of
+                  Just byte -> if not (testBit byte bitIndex) || IS.member n allowedSet then pure True else pure False
+                  Nothing -> pure True
+       in and <$> mapM isAllowedBit [0 .. totalBits - 1]
```

Comment:
`bitsControlCheck` is a pure function and does not need to be monadic
```suggestion
    bitsControlCheck :: [Int] -> Bool
    bitsControlCheck allowedBits =
      let allowedSet = IS.fromList allowedBits
          totalBits = BS.length bs * 8
          isAllowedBit n =
            let byteIndex = n `shiftR` 3
                bitIndex = n .&. 7
             in case BS.indexMaybe bs byteIndex of
                  Just byte -> not (testBit byte bitIndex) || IS.member n allowedSet
                  Nothing -> True
       in all isAllowedBit [0 .. totalBits - 1]
```